server/room-battle.ts

```diff
@@ -1338,6 +1353,367 @@ export class RoomBattle extends RoomGames.RoomGame<RoomBattlePlayer> {
+		if (!options.players && (options.p1 || options.p2)) {
+			options.players = [options.p1?.user, options.p2?.user].map(toID);
+		}
+		for (const userid of options.players || []) {
+			const player = this.makePlayer(userid);
+			this.playerTable[userid] = player;
+			this.room.auth.set(userid, Users.PLAYER_SYMBOL);
+			this[`p${player.num}` as 'p1' | 'p2'] = userid;
+		}
+		process.nextTick(() => this.nextGame());
+	}
+	onConnect(user: User) {
+		const player = this.playerTable[user.id];
+		if (player) {
+			player.sendRoom('|cantleave|');
+		}
+	}
+	makePlayer(userid: string | User): RoomGames.RoomGamePlayer {
+		return new RoomGames.RoomGamePlayer(Users.get(userid) || userid, this, ++this.playerNum);
+	}
+	cleanup() {
+		this.waitingBattle = null;
+		if (this.ready) this.ready = null;
+		if (this.nextBattleTimer) {
+			clearInterval(this.nextBattleTimer);
+			this.nextBattleTimerStart = null;
+		}
+		this.nextBattleTimerStart = null;
+		this.nextBattleTimer = null;
+	}
+	nextGame() {
+		if (this.waitingBattle) {
+			this.waitingBattle.room.add(`Both players are ready! Starting next match!`).update();
+		}
+		this.cleanup();
+
+		const battle = Rooms.createBattle(this.options);
+		if (!battle) throw new Error("Failed to create battle for " + this.title);
+		battle.setParent(this.room);
+		this.games.push({
+			battle,
+			winner: null,
+			rated: battle.rated,
+		});
+		// the absolute result is what counts for rating
+		battle.rated = 0;
+		if (this.needsTimer) {
+			battle.battle?.timer.start();
+		}
+		battle.add(
+			`|html|View the match progress at <a href="/${this.roomid}">${this.roomid}</a>`
+		).update();
+		this.updateDisplay();
+		this.room.add(`|html|<h2>Game ${this.games.length}</h2>`);
+		this.room.add(`|html|<a href="/${battle.roomid}">${battle.title}</a>`);
+		this.room.update();
+	}
+	updateDisplay() {
+		const p1name = this.name(this.p1);
+		const p2name = this.name(this.p2);
+		let buf = Utils.html`<br /><strong>${p1name} and ${p2name}'s Best-of-${this.bestOf} progress:</strong><br />`;
+		for (const userid of [this.p1, this.p2]) {
+			buf += `<span style="text-align: right">`;
+			buf += `${this.name(userid)}: `;
+			for (let i = 0; i < this.bestOf; i++) {
+				if (this.games[i]?.winner === userid) {
+					buf += `<i class="fa fa-circle"></i>`;
+				} else {
+					buf += `<i class="fa fa-circle-o"></i>`;
+				}
+				if ((i + 1) === this.winThreshold) {
+					buf += ` | `;
+				} else {
+					buf += ` `;
+				}
+			}
+			buf += `</span><br />`;
+		}
+		buf += `<br /><br />`;
+		buf += `<table><tr>`;
+
+		for (const userid of [this.p1, null, this.p2]) {
+			if (userid === null) {
+				buf += `<td></td>`;
+				continue;
+			}
+			buf += `<td><center><strong>${this.name(userid)}</strong></center></td>`;
+		}
+
+		buf += `</tr><tr>`;
+
+		for (const userid of [this.p1, null, this.p2]) {
+			if (userid === null) {
+				buf += `<td></td>`;
+				continue;
+			}
+			const name = (Users.get(userid)?.avatar || 'unknownf') + "";
```

Comment:
```suggestion
			let name = Users.get(userid)?.avatar;
			if (!name || typeof name === 'number') name = 'unknownf';
```
The `src` function can't handle numbers even if you convert them to a string.